pdns/common_startup.cc

```diff
@@ -673,12 +675,36 @@ void mainthread()
     sd_notify(0, "READY=1");
 #endif
 
+  const uint32_t secpollInterval = 1800;
+  uint32_t secpollSince = 0;
+  uint32_t domainCacheUpdateSince = 0;
   for(;;) {
-    sleep(1800);
-    try {
-      doSecPoll(false);
+    const uint32_t slept = g_domainCache.getTTL() == 0 ? secpollInterval : std::min(secpollInterval, g_domainCache.getTTL());
+    sleep(slept);  // if any signals arrive, we might run more often than expected.
+
+    domainCacheUpdateSince += slept;
+    if (domainCacheUpdateSince >= g_domainCache.getTTL()) {
+      try {
+        UeberBackend B;
+        B.updateDomainCache();
```

Comment:
There is a race here. Any zones added or removed while this is running are not taken in account. As a result of this a new zone might show up briefly and then disappear for the duration of the cache ttl.
Running with large ttls will be problematic because of this.  For api only systems this will force you to more frequent domain list updates than necessary.

Reply:
Nice catch! I guess one option would be to set a flag during the refresh so that additions happening during that time are also added to a queue, which would be cleared at the end of the update, or something like that.

Reply:
I've added the flag/queue code now.